dashboard/app/models/contact_rollups_processed.rb

```diff
@@ -0,0 +1,94 @@
+# == Schema Information
+#
+# Table name: contact_rollups_processed
+#
+#  id         :integer          not null, primary key
+#  email      :string(255)      not null
+#  data       :json             not null
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+# Indexes
+#
+#  index_contact_rollups_processed_on_email  (email) UNIQUE
+#
+
+class ContactRollupsProcessed < ApplicationRecord
+  self.table_name = 'contact_rollups_processed'
+
+  DEFAULT_BATCH_SIZE = 10000
+
+  # Aggregates data from contact_rollups_raw table and saves the results, one row per email.
+  #
+  # @param [Integer] batch_size number of records to save per INSERT statement.
+  def self.import_from_raw_table(batch_size = DEFAULT_BATCH_SIZE)
+    # Combines data and metadata for each record in contact_rollups_raw table into one JSON field.
+    # The query result has the same number of rows as in contact_rollups_raw.
+    select_query = <<-SQL.squish
+      SELECT email, JSON_OBJECT('sources', sources, 'data', data, 'data_updated_at', data_updated_at) AS data_and_metadata
+      FROM contact_rollups_raw
+    SQL
+
+    # Groups records by emails. Aggregates all data and metadata belong to an email into one JSON field.
+    #
+    # Note: use GROUP_CONCAT instead of JSON_OBJECT_AGG because the current Aurora Mysql version in
+    # production is 5.7.12, while JSON_OBJECT_AGG is only available from 5.7.22.
+    # Because GROUP_CONCAT returns a string, we add a parser function to convert the result to a hash.
+    group_by_query = <<-SQL.squish
+      SELECT email, CONCAT('[', GROUP_CONCAT(data_and_metadata), ']') AS all_data_and_metadata
+      FROM (#{select_query}) AS sub_query
+      GROUP BY email
+    SQL
+
+    # Process the aggregated data row by row and save the results to DB.
+    batch = []
+    ActiveRecord::Base.connection.exec_query(group_by_query).each do |contact|
+      contact_data = parse_contact_data(contact['all_data_and_metadata'])
+
+      processed_contact_data = {}
+      processed_contact_data.merge!(extract_opt_in(contact_data) || {})
+
+      batch << {email: contact['email'], data: processed_contact_data}
+
+      if batch.size == batch_size
+        import! batch, validate: false
```

Comment:
Why no validation? Performance?

Reply:
Yes. The only validation we need is that an email is unique, which will be done at the DB level anyway thanks to unique index on email. I can add an explanation here.